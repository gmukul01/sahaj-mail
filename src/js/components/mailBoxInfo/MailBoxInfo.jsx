import React, { useState, useEffect } from 'react';
import PropTypes from 'prop-types';

import TopBar from './TopBar';
import EmailList from 'components/email/EmailList';
import ActionBar from './ActionBar';

const MailBoxInfo = props => {
	const { selectedEmails, addToSelectedEmails, clearSelectedEmails } = selectedEmailsState([]);
	const { totalEmails, emails, isSentMail, fetchEmails, deleteEmails, readEmails, title } = props;

	useEffect(() => {
		fetchEmails();
	}, []);

	function onRead() {
		readEmails(selectedEmails);
		clearSelectedEmails();
	}

	function onDelete() {
		deleteEmails(selectedEmails);
		clearSelectedEmails();
	}

	function onRefresh() {
		fetchEmails();
		clearSelectedEmails();
	}

	return (
		<section className="content mailBoxInfo">
			<TopBar title={title} totalEmails={totalEmails} />
			<ActionBar {...{ onRefresh, onDelete, onRead }} />
			<EmailList isSentMail={isSentMail} {...{ selectedEmails, addToSelectedEmails, onRead: readEmails, emails }} />
		</section>
	);
};

MailBoxInfo.propTypes = {
	totalEmails: PropTypes.number,
	emails: PropTypes.array,
	fetchEmails: PropTypes.func,
	deleteEmails: PropTypes.func,
	readEmails: PropTypes.func,
	fetchInboxDetails: PropTypes.func
};

export default React.memo(MailBoxInfo);

const selectedEmailsState = initialValue => {
	const [emails, setValue] = useState(initialValue);

	function handleChange(e) {
		const emailId = Number(e.target.value),
			newSelectedEmails = emails.indexOf(emailId) === -1 ? [...emails, emailId] : emails.filter(id => id !== emailId);

		setValue(newSelectedEmails);
	}

	function clearSelectedEmails() {
		setValue([]);
	}

	return { selectedEmails: emails, addToSelectedEmails: handleChange, clearSelectedEmails };
};
