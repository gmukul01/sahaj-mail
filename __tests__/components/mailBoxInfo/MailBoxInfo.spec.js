import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import MailBoxInfo from 'components/mailBoxInfo/MailBoxInfo';

describe('MailBoxInfo Component', () => {
	const mockFetchEmail = jest.fn(),
		mockDeleteEmail = jest.fn(),
		mockReadEmail = jest.fn(),
		initialProps = {
			totalUnread: 2,
			emails: [],
			fetchEmails: mockFetchEmail,
			deleteEmails: mockDeleteEmail,
			readEmails: mockReadEmail,
			title: 'Inbox'
		};

	afterEach(() => {
		mockFetchEmail.mockClear();
		mockDeleteEmail.mockClear();
		mockReadEmail.mockClear();
	});

	it('should match snapshot', () => {
		const mailBoxInfo = renderer.create(<MailBoxInfo {...initialProps} />).toJSON();
		expect(mailBoxInfo).toMatchSnapshot();
	});

	it('should call fetchEmails when refresh is clicked', () => {
		const mailBoxInfo = shallow(<MailBoxInfo {...initialProps} />);
		mailBoxInfo.find('ActionBar').prop('onRefresh')();
		expect(mockFetchEmail.mock.calls.length).toBe(1);
	});

	it('should call deleteEmails when delete is clicked', () => {
		const mailBoxInfo = shallow(<MailBoxInfo {...initialProps} />);
		mailBoxInfo.find('ActionBar').prop('onDelete')();
		expect(mockDeleteEmail.mock.calls.length).toBe(1);
	});

	it('should call onRead when read is clicked', () => {
		const mailBoxInfo = shallow(<MailBoxInfo {...initialProps} />);
		mailBoxInfo.find('ActionBar').prop('onRead')();
		expect(mockReadEmail.mock.calls.length).toBe(1);
	});
});
