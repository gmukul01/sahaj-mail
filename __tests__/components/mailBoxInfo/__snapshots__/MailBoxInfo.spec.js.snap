// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MailBoxInfo Component should match snapshot 1`] = `
<section
  className="content mailBoxInfo"
>
  <div
    className="mailBoxInfo-topbar"
  >
    <div
      className="title"
    >
      Inbox ()
    </div>
  </div>
  <div
    className="mailBoxInfo-actionbar"
  >
    <div
      className="actions"
    >
      <button
        className="button-outlined actions-refresh"
        onClick={[Function]}
      >
        <i
          className="fas fa-sync-alt"
        />
        Refresh
      </button>
      <button
        className="button-outlined"
        onClick={[Function]}
      >
        <i
          className="far fa-eye"
        />
      </button>
      <button
        className="button-outlined"
      >
        <i
          className="fas fa-exclamation"
        />
      </button>
      <button
        className="button-outlined"
        onClick={[Function]}
      >
        <i
          className="far fa-trash-alt"
        />
      </button>
    </div>
    <div>
      <button
        className="button-outlined"
      >
        <i
          className="fas fa-arrow-left"
        />
      </button>
      <button
        className="button-outlined"
      >
        <i
          className="fas fa-arrow-right"
        />
      </button>
    </div>
  </div>
  <div
    className="email-list"
  />
</section>
`;
